posts/2021/08/github-actions-benchmarks.md

```diff
@@ -0,0 +1,336 @@
+process!
+2. The default configuration will run the benchmark in two interleaved passes, with several
+repeats each. This is done to account for the unavoidable deviations from ideality caused
+from co-running processes, as mentioned above.
+3. The statistics for each commit are gathered and a report table is presented. The ratio
+of each test is computed and, if it is greater than a certain threshold (`1.2` by default)
+an error is emitted.
+
+> <sup>†</sup> `asv` supports the notion of a configuration matrix, so you can test your code under
+different environments; e.g. NumPy versions, Python interpreters, etc.
+
+For the benchmark suite of `scikit-image` as of June/July 2021, this ends up taking up to two hours.
+This raises two questions we will answer in the following sections:
+
+1. Although we are trying hard to reduce measurement errors, is that enough? Are these measurements reliable?
+2. Two hours might be too long. Are there any settings we can tune to reduce the runtime
+without reducing the accuracy of the measurements?
+
+# Are CI services reliable enough for benchmarking?
+
+We saw above that CI services are not designed for this kind of tasks, but with some workarounds
+they might be good enough. However, how can be sure? As data-driven scientists, we say let's run
+an experiment!
+
+## The setup
+
+Our experiment is data-drive, so we need to generate some data first. This is our strategy:
+
+* We will benchmark and compare two commits that are exactly the same in terms of tested code.
+* Under ideal conditions, we should see that the performance ratio between the two commits is 1.0.
+In other words, performance should be the same. Of course, these are not ideal conditions, so
+some kind of error is expected. We want to see if it stays reliably under an acceptable threshold.
+* We will implement a GitHub Actions workflow that will run every six hours (four times a day) and
+collect results for a week or more. This will help us account for two things:
+    * Accumulate sufficient data points to answer our question.
+    * Account for time factors like different days of the week (e.g. weekday vs weekend),
+      or the time of the day (3am vs 3pm).
+* The workflow will upload the benchmark results as artifacts we can download and process locally
+  with a Jupyter Notebook.
+
+> Check the [GitHub Actions workflow in this fork](https://github.com/jaimergp/scikit-image/blob/main/.github/workflows/benchmarks-cron.yml)!
+
+## The results
+
+To measure the stability of the different benchmark runs we will be looking at the performance
+ratio of the measurements between the two commits. Since they are identical in terms of tested code,
+ideally they should be all `1.0`. We know this will not happen, but maybe the errors are not that big
+and stay like that regardless the time of the day or the day of the week.
+
+After some days collecting data points, these are the results:
```

Comment:
perhaps specify the actual number of days here

After collecting data for XX days, ...